product-mini/platforms/linux-sgx/enclave-sample/Enclave/Enclave.cpp

```diff
@@ -341,152 +439,223 @@ wasm_runtime_get_module_hash(wasm_module_t module)
+    bool ret = false;
+    wasm_module_inst_t module_inst =
+        (wasm_module_inst_t)gWasmModuleInstMgr.get(wasm_module_inst_idx);
 
-    bh_assert(argc == 1);
+    if (module_inst == nullptr) {
+        ret = false;
+        goto exit;
+    }
 
     wasm_runtime_deinstantiate(module_inst);
 
+    if (!gWasmModuleInstMgr.remove(wasm_module_inst_idx)) {
+        ret = false;
+        goto exit;
+    }
     LOG_VERBOSE("Deinstantiate module success.\n");
+    ret = true;
+exit:
+    return ret;
 }
 
-static void
-handle_cmd_get_exception(uint64 *args, uint32 argc)
+bool
+ecall_handle_cmd_get_exception(uint16_t wasm_module_inst_idx, char *exception,
+                               uint32_t exception_size)
 {
-    uint64 *args_org = args;
-    wasm_module_inst_t module_inst = *(wasm_module_inst_t *)args++;
-    char *exception = *(char **)args++;
-    uint32 exception_size = *(uint32 *)args++;
+    bool ret = false;
+    wasm_module_inst_t module_inst = nullptr;
     const char *exception1;
 
-    bh_assert(argc == 3);
+    if ((exception == nullptr)
+        or (!(module_inst = (wasm_module_inst_t)gWasmModuleInstMgr.get(
+                  wasm_module_inst_idx)))) {
+        ret = false;
+        goto exit;
+    }
 
     if ((exception1 = wasm_runtime_get_exception(module_inst))) {
         snprintf(exception, exception_size, "%s", exception1);
-        args_org[0] = true;
+        ret = true;
     }
     else {
-        args_org[0] = false;
+        ret = false;
     }
+exit:
+    return ret;
 }
 
-static void
-handle_cmd_exec_app_main(uint64 *args, int32 argc)
+bool
+ecall_handle_cmd_exec_app_main(uint16_t wasm_module_inst_idx, char **u_app_argv,
+                               uint32_t app_argc)
 {
-    wasm_module_inst_t module_inst = *(wasm_module_inst_t *)args++;
-    uint32 app_argc = *(uint32 *)args++;
+    bool ret = false;
+    wasm_module_inst_t module_inst = nullptr;
     char **app_argv = NULL;
     uint64 total_size;
     int32 i;
 
-    bh_assert(argc >= 3);
+    if ((u_app_argv == nullptr) or (app_argc == 0)
+        or (!(module_inst = (wasm_module_inst_t)gWasmModuleInstMgr.get(
+                  wasm_module_inst_idx)))) {
+        ret = false;
+        goto exit;
+    }
     bh_assert(app_argc >= 1);
 
     total_size = sizeof(char *) * (app_argc > 2 ? (uint64)app_argc : 2);
 
     if (total_size >= UINT32_MAX
         || !(app_argv = (char **)wasm_runtime_malloc(total_size))) {
         wasm_runtime_set_exception(module_inst, "allocate memory failed.");
-        return;
+        ret = false;
+        goto exit;
     }
 
-    for (i = 0; i < app_argc; i++) {
-        app_argv[i] = (char *)(uintptr_t)args[i];
+    if (!DeepCopyCStrArray(u_app_argv, app_argv, app_argc)) {
+        wasm_runtime_free(app_argv);
+        app_argv = nullptr;
+        ret = false;
+        goto exit;
     }
 
     wasm_application_execute_main(module_inst, app_argc - 1, app_argv + 1);
 
+    for (uint32 i = 0; i < app_argc; i++) {
```

Comment:
The return value may be written back into app_argv[0..1], had better copy them back into u_app_argv

Reply:
OK

Reply:
It need some time

Reply:
> The return value may be written back into app_argv[0..1], had better copy them back into u_app_argv

Hello I want to confirm, it means 1. put the return value (`bool` type) into app_argv[0..1] or 2. content in `app_argv` can be modified by wasm_application_execute_main, and you want this modification propagates to outside of Enclave

Reply:
Hi, I checked again, wasm_application_execute_main puts the return value into `*(int *)argv`, here should be `*(int *)(app_argv + 1)`, see:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/include/wasm_export.h#L836-L838

So my understanding is that we had better write`*(int *)(app_argv + 1)` back into `*(int *)u_app_argv`?

Reply:
I got it, so it's case 2, and I have to let the content in `*(int*)argv` propagates to outside Enclave